executor/analyze.go

```diff
@@ -496,22 +527,458 @@ type AnalyzeFastTask struct {
+	for {
+		loc, ok := <-e.sampLocs
+		if !ok {
+			return
+		}
+		req := &tikvrpc.Request{
+			Type: tikvrpc.CmdDebugGetRegionProperties,
+			DebugGetRegionProperties: &debugpb.GetRegionPropertiesRequest{
+				RegionId: loc.Region.GetID(),
+			},
+		}
+		var resp *tikvrpc.Response
+		var rpcCtx *tikv.RPCContext
+		rpcCtx, *err = e.cache.GetRPCContext(bo, loc.Region)
+		if *err != nil {
+			return
+		}
+		ctx := context.Background()
+		resp, *err = client.SendRequest(ctx, rpcCtx.Addr, req, tikv.ReadTimeoutMedium)
+		if *err != nil {
+			return
+		}
+		if resp.DebugGetRegionProperties == nil || len(resp.DebugGetRegionProperties.Props) == 0 {
+			*needRebuild = true
+			return
+		}
+		for _, prop := range resp.DebugGetRegionProperties.Props {
+			if prop.Name == "mvcc.num_rows" || prop.Name == "num_rows" {
+				var cnt uint64
+				cnt, *err = strconv.ParseUint(prop.Value, 10, 64)
+				if *err != nil {
+					return
+				}
+				newCount := atomic.AddUint64(&e.rowCount, cnt)
+				task := &AnalyzeFastTask{
+					Location:  loc,
+					LRowCount: newCount - cnt,
+					RRowCount: newCount,
+				}
+				*sampTasks = append(*sampTasks, task)
+			}
+		}
+	}
+}
+
+// buildSampTask return tow variable, the first bool is whether the task meeting region error
+// and need to rebuild.
+func (e *AnalyzeFastExec) buildSampTask() (bool, error) {
+	// Do get regions row count.
+	bo := tikv.NewBackoffer(context.Background(), 500)
+	atomic.StoreUint64(&e.rowCount, 0)
+	needRebuildForRoutine := make([]bool, e.concurrency)
+	errs := make([]error, e.concurrency)
+	sampTasksForRoutine := make([][]*AnalyzeFastTask, e.concurrency)
+	e.sampLocs = make(chan *tikv.KeyLocation, e.concurrency)
+	e.wg.Add(e.concurrency)
+	for i := 0; i < e.concurrency; i++ {
+		go e.getSampRegionsRowCount(bo, &needRebuildForRoutine[i], &errs[i], &sampTasksForRoutine[i])
+	}
+
+	store, _ := e.ctx.GetStore().(tikv.Storage)
+	e.cache = store.GetRegionCache()
+	startKey, endKey := tablecodec.GetTableHandleKeyRange(e.physicalTableID)
+	var loc *tikv.KeyLocation
+	var err error
+	// extract all regions contain the table
+	e.scanTasks = e.scanTasks[:0]
+	e.sampTasks = e.sampTasks[:0]
+	// If length of KeyLocation.StartKey is 0, then the start key is infinitesimal.
+	// If length of KeyLocation.EndKey is 0, then the end kety is infinity.
+	for loc, err = e.cache.LocateKey(bo, startKey); bytes.Compare(loc.StartKey, endKey) <= 0 && err == nil; loc, err = e.cache.LocateKey(bo, loc.EndKey) {
+		if bytes.Compare(endKey, loc.EndKey) < 0 || bytes.Compare(loc.StartKey, startKey) < 0 || len(loc.StartKey) == 0 || len(loc.EndKey) == 0 {
+			e.scanTasks = append(e.scanTasks, loc)
+			if len(loc.StartKey) == 0 || bytes.Compare(loc.StartKey, startKey) < 0 {
+				loc.StartKey = startKey
+			}
+			if len(loc.EndKey) == 0 || bytes.Compare(endKey, loc.EndKey) < 0 {
+				loc.EndKey = endKey
+				break
+			}
+		} else {
+			e.sampLocs <- loc
+		}
+	}
+	if err != nil {
+		return false, err
+	}
+	close(e.sampLocs)
```

Comment:
we should close this channel before we returning an error.

Reply:
Now there's no error check here.